Issue #3283375 by SV: As a LU I want to disable following me for other users

volodymyr-sydor opened the pull request:
## Problem
As a LU I want to disable user following

## Solution
- Added possibility to disable and prevent following for other users

## Issue tracker
- https://www.drupal.org/project/social/issues/3283375
- https://getopensocial.atlassian.net/browse/YANG-7374

## How to test
- [ ] Using the latest version of Open Social with the social_follow_user module enabled
- [ ] As a SM, activate `/admin/config/opensocial/follow-user` user follow
- [ ] As a LU, try to enable `/user/ID/edit` the `Prevent members from following me` option
- [ ] Switch back to SM and go to the info page of the LU
- [ ] When opened page, the follow button is not available anymore

## Definition of done
### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] If applicable (I.E. new hook_updates) the update path from previous versions (major and minor versions) are tested
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Screenshots
![Screenshot-2022-05-31-at-16-44-03-png-2740×1634-](https://user-images.githubusercontent.com/25609390/171190029-8644a8d5-8211-4ece-bb5c-9fc5aa786e57.png)

![Screenshot-2022-05-31-at-16-53-32-png-2388×1576-](https://user-images.githubusercontent.com/25609390/171190738-02b7075d-7e53-4b55-9ed7-7db82458d03b.png)

## Release notes
Provides an ability to disable follow feature for other users

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

ribel commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -151,7 +154,14 @@ function social_follow_user_theme(): array {
+
+  $view_mode = $variables['elements']['#view_mode'];
+
+  if ($view_mode === 'teaser') {
```
@volodymyr-sydor Do we really need to add conditions for specific view modes?
Isn't it better to check for all possible view modes?

volodymyr-sydor replied:
We use follow functionality in specific view modes and if we add it then it will be available

ribel replied:
Ok

ribel commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -162,9 +172,23 @@ function social_follow_user_preprocess_profile(array &$variables): void {
         $variables['following'] = TRUE;
```
@volodymyr-sydor I see we already have this `following` variable. Where it is used?
Do we really need extra `disable_following` variable?

volodymyr-sydor replied:
Yes, because followers/following variables uses to indicate which label we should use on the different followers/following list
but in current case we should have addition flag to check follow privacy rule

ribel replied:
Ok

ribel commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -187,3 +211,73 @@ function social_follow_user_preprocess_flag(array &$variables): void {
+
+  if ($profile instanceof ProfileInterface && $active) {
+    $user_data = \Drupal::service('user.data');
+    $disable_following = $user_data->get('social_profile_privacy', (int) $profile->get('uid')->target_id, 'disable_following');
```
@volodymyr-sydor I would suggest to rename it to `following_enabled`
 or `following_enabled`.

Double negatives are bad in plain English as well as computer code. See explanation [here](https://stackoverflow.com/questions/13470724/naming-convention-isenable-vs-isdisable).

volodymyr-sydor replied:
I agree but then using it for this label seems doesn't match:
<img width="902" alt="Screenshot 2022-06-03 at 09 23 54" src="https://user-images.githubusercontent.com/25609390/171798888-b8ef9658-3475-4abf-9bd4-8dbbe272cd1a.png">

volodymyr-sydor replied:
done

ribel reviewed (changes requested):
@volodymyr-sydor I have tested functionality and all works good and as expacted.
But I have suggestion for naming. I know it started from ticket title, but in code I think it would be better to fix it before it merged.

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

volodymyr-sydor commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -187,3 +211,73 @@ function social_follow_user_preprocess_flag(array &$variables): void {
+ * @param array $variables
+ *   Array with variables of a profile.
+ */
+function _social_follow_user_disable_following(array &$variables): void {
```
done

ribel commented on modules/social_features/social_follow_user/social_follow_user.install:
```diff
@@ -149,3 +156,28 @@ function social_follow_user_update_11203(): string {
+/**
+ * Enable following status of all existing users by default.
+ */
+function social_follow_user_update_11402(): void {
```
@volodymyr-sydor It would be good to avoid this update hook, since ability to follow should be enabled by default for all users. Also on platforms where there are 50k of users it may such update may out of memory, so we would need to do it with batch.

ribel commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -187,3 +212,80 @@ function social_follow_user_preprocess_flag(array &$variables): void {
+ *
+ * @throws \Drupal\Core\Entity\EntityStorageException
+ */
+function social_follow_user_user_insert(UserInterface $user): void {
```
@volodymyr-sydor It would be good to avoid this update hook, since ability to follow should be enabled by default for all users.

ribel commented on modules/social_features/social_follow_user/src/Service/SocialFollowUserHelperInterface.php:
```diff
@@ -0,0 +1,67 @@
+   * @return bool
+   *   TRUE or FALSE depending upon following status.
+   */
+  public function isFollowingEnabled(ProfileInterface $profile): bool;
```
@volodymyr-sydor Do we really need to have this method? Can we just have `setFollowingStatus()` and `getFollowingStatus()`?

ribel commented on modules/social_features/social_follow_user/src/Service/SocialFollowUserHelper.php:
```diff
@@ -0,0 +1,104 @@
+      // Check if user already followed.
+      /** @var \Drupal\flag\FlagInterface $flag */
+      $flag = $this->flagService->getFlagById('follow_user');
+      // And display only "Unfollow" button because we should leave the ability
```
@volodymyr-sydor Do we have it in acceptance criteria that other users should be able to unfollow if following is disabled at all for this user?

ribel reviewed (changes requested):
@volodymyr-sydor I still have some questions - please check.
Also please move PR from Draft and merge main branch to fix failing tests due to not applied patch.

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

ribel commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -187,3 +211,90 @@ function social_follow_user_preprocess_flag(array &$variables): void {
+      '#type' => 'checkbox',
+      '#title' => t('Allow members to follow me'),
+      '#default_value' => $following_enabled,
+      '#description' => t('This section allows you to turn off the ability for 
```
@volodymyr-sydor Please change to: _This selection allows you to turn **on** the ability for members to follow your activity and receive notifications about new content you create._

volodymyr-sydor replied:
done

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

Ressinel commented on modules/social_features/social_follow_user/src/Service/SocialFollowUserHelper.php:
```diff
@@ -0,0 +1,104 @@
+   * {@inheritdoc}
+   */
+  public function setFollowingStatus(int $uid, $status = TRUE): void {
+    $this->userData->set(
```
@volodymyr-sydor I more like an approach with a new profile field instead of the user data. Since if we use the user data we need to use batch for hook_install to set the default value and I think it is not good https://github.com/goalgorilla/open_social/pull/2967/files#diff-d3d7b1d9964a18dba55138aef6c9585a65978162753ff93181c2709491e32289R38
But I also understand that many fields for the profile are also not well. So I want to hear your thoughts @navneet0693 @tbsiqueira

cc @ribel

volodymyr-sydor replied:
@Ressinel Yes, seems like it will be even easier if we create a new profile field, for example, `field_profile_following` then we don't need to update existing users & add any action for new accounts, etc.
@navneet0693 @tbsiqueira are you ok with that approach?

volodymyr-sydor replied:
@tbsiqueira 
The acceptance criteria are that `everyone that is already following the user should stay a follower` 
> Will my account still be listed under the "following users" overview from another user? 

Yes, because currently, we disabling only the possibility of following a user. 

If we need to extend functionality and hide followers from the list or etc I would recommend do it separately in follow-up and decide which AC should be applied, because it could make the process review of the task more complicated than was planned. Also, it will require additional testing.  
cc: @ribel

volodymyr-sydor commented:
> @volodymyr-sydor please also check the user profile page itself, even after I disable the "follow" option in one user, I'm still able to follow it from the "profile" page.
> 
> ![image](https://user-images.githubusercontent.com/11061892/174040103-48b89f29-d060-43f0-bb48-47927de9a04c.png)

@tbsiqueira 
The button displays in the statistic block, have you applied a theme patch as well? (here is the [issue](https://www.drupal.org/project/socialblue/issues/3283531))

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

ribel commented:
Hey @Ressinel @tbsiqueira, @volodymyr-sydor implemented requested changes and it looks good to me now. 
Could you check it again and let us know if we can merged it?

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

Ressinel reviewed (commented):
@volodymyr-sydor I have found a case, where the user can follow another one even if the setting `Allow members to follow me`, is disabled:
- `user_1` and `user_2`
- `user_1` has followed `user_2`
- `user_2` set `FALSE` for the `Allow members to follow me` setting
- `user_1` is still able to unfollow `user_2`
- `user_1` press unfollow button and at the same time can follow `user_2` again even if `Allow members to follow me` is `FALSE` for `user_2`

Not sure that it is critical but at least it is not logical that some users can re-follow even if `Allow members to follow me` is `FALSE`.

volodymyr-sydor commented:
> @volodymyr-sydor I have found a case, where the user can follow another one even if the setting `Allow members to follow me`, is disabled:
> 
> * `user_1` and `user_2`
> * `user_1` has followed `user_2`
> * `user_2` set `FALSE` for the `Allow members to follow me` setting
> * `user_1` is still able to unfollow `user_2`
> * `user_1` press unfollow button and at the same time can follow `user_2` again even if `Allow members to follow me` is `FALSE` for `user_2`
> 
> Not sure that it is critical but at least it is not logical that some users can re-follow even if `Allow members to follow me` is `FALSE`.

@Ressinel 
We've left the possibility to unfollow the user. If it's caused a new issue I would suggest do not show "unfollow" button as well. So, I will remove it as it even wasn't in the acceptance criteria

Ressinel commented:
> > @volodymyr-sydor I have found a case, where the user can follow another one even if the setting `Allow members to follow me`, is disabled:
> > 
> > * `user_1` and `user_2`
> > * `user_1` has followed `user_2`
> > * `user_2` set `FALSE` for the `Allow members to follow me` setting
> > * `user_1` is still able to unfollow `user_2`
> > * `user_1` press unfollow button and at the same time can follow `user_2` again even if `Allow members to follow me` is `FALSE` for `user_2`
> > 
> > Not sure that it is critical but at least it is not logical that some users can re-follow even if `Allow members to follow me` is `FALSE`.
> 
> @Ressinel We've left the possibility to unfollow the user. If it's caused a new issue I would suggest do not show "unfollow" button as well. So, I will remove it as it even wasn't in the acceptance criteria

@volodymyr-sydor I think we should leave the possibility to unfollow in this case. At least it sounds logical.
@tbsiqueira @ribel need your thoughts as well

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2967-xi5hc1yplc489g6jlevpvqkfzhlujwx8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62961f196eb9e7df1a3aef01

volodymyr-sydor commented:
> @Ressinel As @volodymyr-sydor said adding possibility to unfollow even if you disable follow was already out of the scope of this story. But since it also looked logical to us we implemented it. But then we come to the use case which you described. And I agree that it would be good to fix that, but not sure how much time it will take and we already spent more time that planned on it.
> 
> @volodymyr-sydor maybe you can still investigate it a bit and besides that we also now need to fix conflict in composer after recent merges.

I've adjusted the access check and now after "unfollow" action the button is not available for user with disabled "following" option

volodymyr-sydor commented on modules/social_features/social_follow_user/social_follow_user.module:
```diff
@@ -162,7 +172,21 @@ function social_follow_user_preprocess_profile(array &$variables): void {
+ * Implements hook_preprocess_HOOK().
+ */
+function social_follow_user_preprocess_profile__profile__statistic(array &$variables): void {
+  if (!\Drupal::config('social_follow_user.settings')->get('status')) {
```
done

Ressinel reviewed (approved):
Great work! Looks good to me!

Ressinel commented:
🍒-picked to `11.4.x`
